### PR TITLE
To distinguish BOM update and dependency update in  Global Upper Bounds Fixes

### LINKED ISSUE #568: Dashboard can suggest a global upper bounds fix for an artifact itself:
This is interesting. Not sure whether we should filter this out completely or replace it with a message that the BOM is out of date.

Global Upper Bounds Fixes

Suggested updates to bring this artifact into sync with the highest versions of its dependencies used by any artifact in the BOM:

    Upgrade com.google.protobuf:protobuf-java:jar:3.6.1 to com.google.protobuf:protobuf-java:jar:3.7.1:

    Add this dependency element to the pom.xml for com.google.protobuf:protobuf-java:3.6.1:

    <dependency>
      <groupId>com.google.protobuf</groupId>
      <artifactId>protobuf-java</artifactId>
      <version>3.7.1</version>
    </dependency>

If the pom.xml for com.google.protobuf:protobuf-java:3.6.1 already includes this dependency, update the version of the existing dependency element. Otherwise add a new dependency element to the dependencyManagement section.

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #568.

Screenshot:

image

### CODE REVIEW

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 1]
```diff
@@ -326,9 +326,7 @@ private static ArtifactResults generateArtifactReport(
                   .equals(Artifacts.toCoordinates(dependencyPath.get(0)))));
 
       Map<String, Object> templateData = new HashMap<>();
-      templateData.put("groupId", artifact.getGroupId());
-      templateData.put("artifactId", artifact.getArtifactId());
-      templateData.put("version", artifact.getVersion());
```
Comment (suztomo) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: These three variables are now defined in component.ftl using `artifact`.

File: dashboard/src/main/resources/templates/component.ftl [hunk 2]
```diff
@@ -24,21 +27,42 @@
...
         <#list globalUpperBoundFailures as lower, upper>
-          <li>Upgrade ${lower} to ${upper}:
-          
-          <p>Add this dependency element to the pom.xml for ${groupId}:${artifactId}:${version}:</p>
-          
-<pre class="suggested-dependency-mediation"><code>&lt;dependency>
-  &lt;groupId>${upper.getGroupId()}&lt;/groupId>
-  &lt;artifactId>${upper.getArtifactId()}&lt;/artifactId>
-  &lt;version>${upper.getVersion()}&lt;/version>
+          <#if lower == artifact>
```
Comment (suztomo) [inline dashboard/src/main/resources/templates/component.ftl]: Now it distinguish whether the upgrade is the BOM member or not.

File: dashboard/src/main/resources/templates/component.ftl [hunk 3]
```diff
@@ -1,14 +1,17 @@
 <!DOCTYPE html>
 <html lang="en-US">
   <#include "macros.ftl">
+  <#assign groupId = artifact.getGroupId()>
+  <#assign artifactId = artifact.getArtifactId()>
+  <#assign version = artifact.getVersion()>
   <head>
     <meta charset="utf-8" />
-    <title>Google Cloud Platform Dependency Analysis Report for ${groupId}:${artifactId}:${version}</title>
+    <title>Google Cloud Platform Dependency Analysis Report for ${artifact}</title>
```
Comment (elharo) [inline dashboard/src/main/resources/templates/component.ftl]: revert; it's unclear what toString does here

File: dashboard/src/main/resources/templates/component.ftl [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
     <link rel="stylesheet" href="dashboard.css" />
     <script src="dashboard.js"></script>
   </head>
   <body>
-    <h1>Dependency Analysis of ${groupId}:${artifactId}:${version}</h1>
+    <h1>Dependency Analysis of ${artifact}</h1>
```
Comment (elharo) [inline dashboard/src/main/resources/templates/component.ftl]: ditto; weird toString output has been a problem lately

File: dashboard/src/main/resources/templates/component.ftl (hunk 2 above)
Comment (elharo) [inline dashboard/src/main/resources/templates/component.ftl]: What's the equals semantics here? identity or equality?

File: dashboard/src/main/resources/templates/component.ftl [hunk 5]
```diff
@@ -24,21 +27,42 @@
...
-          
-          </li>
+
+            </li>
+          </#if>
         </#list>
       </ul>
       
-      <p>If the pom.xml for ${groupId}:${artifactId}:${version} already includes this dependency,
+      <p>If the pom.xml for ${artifact} already includes this dependency,
```
Comment (elharo) [inline dashboard/src/main/resources/templates/component.ftl]: ditto, no toString

File: dashboard/src/main/resources/templates/component.ftl (hunk 3 above)
Comment (suztomo) [reply dashboard/src/main/resources/templates/component.ftl]: Makes sense. Reverted.

File: dashboard/src/main/resources/templates/component.ftl (hunk 4 above)
Comment (suztomo) [reply dashboard/src/main/resources/templates/component.ftl]: Reverted.

File: dashboard/src/main/resources/templates/component.ftl (hunk 5 above)
Comment (suztomo) [reply dashboard/src/main/resources/templates/component.ftl]: Reverted.

File: dashboard/src/main/resources/templates/component.ftl (hunk 2 above)
Comment (suztomo) [reply dashboard/src/main/resources/templates/component.ftl]: It's equality ([Freemarker Expressions - Comparison](https://freemarker.apache.org/docs/dgui_template_exp.html#dgui_template_exp_comparison)). But I updated the code to compare only the coordinates, without 'file' attribute of Artifact instance.

**Comment by suztomo** [comment]
Updated; the sentence should be within the list element. Updated the screenshot of this PR.
PTAL.

File: example-problems/no-such-method-error-signature-mismatch/pom.xml [hunk 6]
```diff
@@ -39,6 +39,34 @@
           <mainClass>io.grpc.internal.App</mainClass>
         </configuration>
       </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
```
Comment (elharo) [inline example-problems/no-such-method-error-signature-mismatch/pom.xml]: This change doesn't seem related to this PR.

Comment (suztomo) [reply example-problems/no-such-method-error-signature-mismatch/pom.xml]: Sorry, missed git operation. Removed.

**Comment by suztomo** [comment]
I see "TOOL ABORTED" message in Fusion. Going to ask Kokoro team.

**Pull request merged**